File: D_Signs_of_Prefix_Sums.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool check(int n, const string& s, int K) {
    int L = 0, R = 0;
    for (int i = 0; i < n; ++i) {
        int nL = L - K;
        int nR = R + K;
        
        if (s[i] == '+') {
            nL = max(nL, 1);
            if (L == R && L >= 1 && L <= R) {
            }
        } else if (s[i] == '-') {
            nR = min(nR, -1);
        } else {
            nL = max(nL, 0);
            nR = min(nR, 0);
            if (L == 0 && R == 0) return false;
        }

        if (nL > nR) return false;
        L = nL;
        R = nR;
    }
    return true;
}

void solve() {
    int n;
    cin >> n;
    string s;
    cin >> s;

    if (s[0] == '0') {
        cout << -1 << "\n";
        return;
    }
    for (int i = 0; i + 1 < n; ++i) {
        if (s[i] == '0' && s[i + 1] == '0') {
            cout << -1 << "\n";
            return;
        }
    }

    int low = 1, high = n + 1, ans = -1;
    while (low <= high) {
        int mid = low + (high - low) / 2;
        if (check(n, s, mid)) {
            ans = mid;
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }
    cout << ans << "\n";
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    int t;
    if (cin >> t) {
        while (t--) {
            solve();
        }
    }
    return 0;
}

File: D_Signs_of_Prefix_Sums.py
import sys


def solve():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return

    qrTNum = int(data[0])
    idx = 1
    out = []

    for _ in range(qrTNum):
        n = int(data[idx])
        s = data[idx + 1]
        idx += 2

        if s[0] == "0" or "00" in s:
            out.append("-1")
            continue

        low = 1
        high = n + 1
        ans = -1

        while low <= high:
            mid = (low + high) // 2

            L = 0
            R = 0
            possible = True

            for ch in s:
                if L == R:
                    x = L
                    if ch == "+":
                        l1, r1 = max(1, x - mid), x - 1
                        l2, r2 = max(1, x + 1), x + mid

                        valid1 = l1 <= r1
                        valid2 = l2 <= r2

                        if valid1 and valid2:
                            nL = min(l1, l2)
                            nR = max(r1, r2)
                        elif valid1:
                            nL, nR = l1, r1
                        elif valid2:
                            nL, nR = l2, r2
                        else:
                            possible = False
                            break
                    elif ch == "-":
                        l1, r1 = x - mid, min(-1, x - 1)
                        l2, r2 = x + 1, min(-1, x + mid)

                        valid1 = l1 <= r1
                        valid2 = l2 <= r2

                        if valid1 and valid2:
                            nL = min(l1, l2)
                            nR = max(r1, r2)
                        elif valid1:
                            nL, nR = l1, r1
                        elif valid2:
                            nL, nR = l2, r2
                        else:
                            possible = False
                            break
                    else:
                        if x != 0 and abs(x) <= mid:
                            nL, nR = 0, 0
                        else:
                            possible = False
                            break
                else:
                    rawL = L - mid
                    rawR = R + mid
                    if ch == "+":
                        nL = max(1, rawL)
                        nR = rawR
                    elif ch == "-":
                        nL = rawL
                        nR = min(-1, rawR)
                    else:
                        if rawL <= 0 <= rawR:
                            nL, nR = 0, 0
                        else:
                            possible = False
                            break

                if nL > nR:
                    possible = False
                    break

                L, R = nL, nR

            if possible:
                ans = mid
                high = mid - 1
            else:
                low = mid + 1

        out.append(str(ans))

    sys.stdout.write("\n".join(out) + "\n")


if __name__ == "__main__":
    solve()